src/util/cilfacade.ml

```diff
@@ -80,6 +80,51 @@ let end_basic_blocks f =
   let thisVisitor = new allBBVisitor in
   visitCilFileSameGlobals thisVisitor f
 
+class removeLabelsVisitor = object
+  inherit nopCilVisitor
+
+  val nestingDepth = ref 0
+
+  method! vstmt s =
+    let new_labels = List.filter_map (function Label(str,loc,b) -> None | x -> Some x) s.labels in
+    match s.skind with
+    | Loop _ -> nestingDepth := !nestingDepth+1; ChangeDoChildrenPost({s with labels = new_labels}, fun x -> nestingDepth := !nestingDepth-1; x)
+    | Continue loc -> if !nestingDepth = 0 then
+        ChangeTo({s with labels = new_labels; skind = Break loc})
+      else
+        ChangeTo({s with labels = new_labels})
+    | _ -> ChangeDoChildrenPost({s with labels = new_labels}, Fun.id)
+end
+
+class loopUnrollingVisitor = object
+  (* Labels are simply handled by removing them. It is always correct to simply jump into the non-unrolled part of the loop!          *)
+  (* The price we need to pay for this simpler handling of labels and still wanting to benefit from unrolling nested loops is that    *)
+  (* we need to put our unrolled code into fake while(1) loops such that breaks have something to break out of, add artificial breaks *)
+  (* at the end of loops, and replacing (top level!) continues with breaks. *)
+  (* This is still a lot more intuitive than complicated handling of gotos as would be required when doing it after the removal of breaks  *)
+  inherit nopCilVisitor
+    
+  method! vstmt s =
+    let labelsRemover = new removeLabelsVisitor in
+    match s.skind with
+    | Loop (b,loc, loc2, break , continue) ->
+      let duplicate_and_rem_labels s =
+        match s.skind with
+        | Loop (b,loc, loc2, break , continue) ->
+          let s = { s with sid = s.sid } in
+          let factor = GobConfig.get_int "exp.unrolling-factor" in
+          let bstmts = b.bstmts @ [mkStmt @@ Break loc] in
+          let copies = List.init (factor) (fun _ -> visitCilStmt labelsRemover (mkStmt (Loop (mkBlock bstmts,loc,loc2,None,None)))) in
```

Comment:
This calls `labelsRemover` `factor` times, instead of doing it just once. Could just do with `BatList.make`.

Reply:
Yes, and this is by design! We need fresh copies of the loop body, otherwise we run into issues with sharing again.

Reply:
Oh, so `removeLabelsVisitor` implicitly does a lot more than just remove labels. It also copies the entire statement recursively. It's not just that `{s with labels = new_labels}` in the default case there is used to remove the labels, but in all the cases where there are no labels, this still guarantees a fresh physical instance of the record.

I think all this physical instance trickery should be explicitly commented because it's key to all this working but very easy to accidentally want to optimize away.